https://www.kaggle.com/datasets/patrickb1912/ipl-complete-dataset-20082020?select=deliveries.csv

In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import numpy as np
from scipy.stats import gaussian_kde
from plotly.subplots import make_subplots
import pprint

In [74]:
data = pd.read_csv("../datasets/deliveries.csv")

In [3]:
data.shape

(260920, 17)

In [4]:
#### to dispaly all feature without wrap in data.head()
pd.set_option('display.max_columns', None)

In [5]:
data.head()

match_id  inning           batting_team                 bowling_team  over  \
0    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
1    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
2    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
3    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   
4    335982       1  Kolkata Knight Riders  Royal Challengers Bangalore     0   

   ball       batter   bowler  non_striker  batsman_runs  extra_runs  \
0     1   SC Ganguly  P Kumar  BB McCullum             0           1   
1     2  BB McCullum  P Kumar   SC Ganguly             0           0   
2     3  BB McCullum  P Kumar   SC Ganguly             0           1   
3     4  BB McCullum  P Kumar   SC Ganguly             0           0   
4     5  BB McCullum  P Kumar   SC Ganguly             0           0   

   total_runs extras_type  is_wicket player_dismissed dismissal_kind fielder  
0           1     legbyes          0              NaN            NaN     NaN  
1           0         NaN          0              NaN            NaN     NaN  
2           1       wides          0              NaN            NaN     NaN  
3           0         NaN          0              NaN            NaN     NaN  
4           0         NaN          0              NaN            NaN     NaN

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260920 entries, 0 to 260919
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          260920 non-null  int64 
 1   inning            260920 non-null  int64 
 2   batting_team      260920 non-null  object
 3   bowling_team      260920 non-null  object
 4   over              260920 non-null  int64 
 5   ball              260920 non-null  int64 
 6   batter            260920 non-null  object
 7   bowler            260920 non-null  object
 8   non_striker       260920 non-null  object
 9   batsman_runs      260920 non-null  int64 
 10  extra_runs        260920 non-null  int64 
 11  total_runs        260920 non-null  int64 
 12  extras_type       14125 non-null   object
 13  is_wicket         260920 non-null  int64 
 14  player_dismissed  12950 non-null   object
 15  dismissal_kind    12950 non-null   object
 16  fielder           9354 non-null    obj

In [7]:
data.describe()

match_id         inning           over           ball  \
count  2.609200e+05  260920.000000  260920.000000  260920.000000   
mean   9.070665e+05       1.483531       9.197677       3.624486   
std    3.679913e+05       0.502643       5.683484       1.814920   
min    3.359820e+05       1.000000       0.000000       1.000000   
25%    5.483340e+05       1.000000       4.000000       2.000000   
50%    9.809670e+05       1.000000       9.000000       4.000000   
75%    1.254066e+06       2.000000      14.000000       5.000000   
max    1.426312e+06       6.000000      19.000000      11.000000   

        batsman_runs     extra_runs     total_runs      is_wicket  
count  260920.000000  260920.000000  260920.000000  260920.000000  
mean        1.265001       0.067806       1.332807       0.049632  
std         1.639298       0.343265       1.626416       0.217184  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000       0.000000  
50%         1.000000       0.000000       1.000000       0.000000  
75%         1.000000       0.000000       1.000000       0.000000  
max         6.000000       7.000000       7.000000       1.000000

In [8]:
def Columns_Info(data):
    """
    Display a table showing the number of unique values for each column in the DataFrame.
    """
    unique_counts = {
        "Column": data.columns,
        "Unique Values": [data[col].nunique() for col in data.columns],
        'null values': [data[col].isna().sum() for col in data.columns]
    }
    
    result_df = pd.DataFrame(unique_counts)
    display(result_df)
Columns_Info(data)

Column  Unique Values  null values
0           match_id           1095            0
1             inning              6            0
2       batting_team             19            0
3       bowling_team             19            0
4               over             20            0
5               ball             11            0
6             batter            673            0
7             bowler            530            0
8        non_striker            663            0
9       batsman_runs              7            0
10        extra_runs              7            0
11        total_runs              8            0
12       extras_type              5       246795
13         is_wicket              2            0
14  player_dismissed            629       247970
15    dismissal_kind             10       247970
16           fielder            607       251566

## All Unique Values

In [14]:
data['inning'].value_counts()

inning
1    135018
2    125741
3        77
4        72
5         8
6         4
Name: count, dtype: int64

In [ ]:
data['batting_team'].value_counts()  ## bowling team is same as bating team

batting_team
Mumbai Indians                 31437
Kolkata Knight Riders          29514
Chennai Super Kings            28651
Royal Challengers Bangalore    28205
Rajasthan Royals               26242
Kings XI Punjab                22646
Sunrisers Hyderabad            21843
Delhi Daredevils               18786
Delhi Capitals                 10946
Deccan Chargers                 9034
Punjab Kings                    6833
Gujarat Titans                  5494
Pune Warriors                   5443
Lucknow Super Giants            5400
Gujarat Lions                   3566
Rising Pune Supergiant          1900
Royal Challengers Bengaluru     1818
Kochi Tuskers Kerala            1582
Rising Pune Supergiants         1580
Name: count, dtype: int64

In [16]:
data['batsman_runs'].value_counts()

batsman_runs
0    103940
1     96778
4     29850
2     16453
6     13051
3       783
5        65
Name: count, dtype: int64

In [17]:
data['extra_runs'].value_counts()

extra_runs
0    246795
1     12628
2       585
4       504
5       325
3        82
7         1
Name: count, dtype: int64

In [18]:
data['total_runs'].value_counts()

total_runs
1    108440
0     90438
4     30221
2     17323
6     12964
3       922
5       524
7        88
Name: count, dtype: int64

In [19]:
data['extras_type'].value_counts()

extras_type
wides      8380
legbyes    4001
noballs    1069
byes        673
penalty       2
Name: count, dtype: int64

In [20]:
data['dismissal_kind'].value_counts()

dismissal_kind
caught                   8063
bowled                   2212
run out                  1114
lbw                       800
caught and bowled         367
stumped                   358
retired hurt               15
hit wicket                 15
obstructing the field       3
retired out                 3
Name: count, dtype: int64

In [23]:

data['fielder'].value_counts().head(30)

fielder
MS Dhoni          220
KD Karthik        199
AB de Villiers    141
RV Uthappa        133
V Kohli           131
WP Saha           131
SK Raina          119
RA Jadeja         116
SV Samson         115
RG Sharma         111
KA Pollard        106
S Dhawan          102
RR Pant           100
MK Pandey          98
DA Warner          97
PA Patel           95
Q de Kock          90
KL Rahul           89
F du Plessis       87
DA Miller          86
NV Ojha            82
DJ Bravo           81
AM Rahane          81
AC Gilchrist       75
AT Rayudu          75
AR Patel           72
SA Yadav           70
HH Pandya          67
Ishan Kishan       66
SPD Smith          64
Name: count, dtype: int64

extras_type	5	246795  
13	is_wicket	2	0                 
14	player_dismissed	629	247970                
15	dismissal_kind	10	247970               
16	fielder	607	251566         


In [31]:
Columns_Info(data)

Column  Unique Values  null values
0           match_id           1095            0
1             inning              6            0
2       batting_team             19            0
3       bowling_team             19            0
4               over             20            0
5               ball             11            0
6             batter            673            0
7             bowler            530            0
8        non_striker            663            0
9       batsman_runs              7            0
10        extra_runs              7            0
11        total_runs              8            0
12       extras_type              1            0
13         is_wicket              2            0
14  player_dismissed            630            0
15    dismissal_kind             11            0
16           fielder            608            0

In [44]:
data['batting_team'].value_counts().head()

batting_team
Mumbai Indians                 31437
Kolkata Knight Riders          29514
Chennai Super Kings            28651
Royal Challengers Bangalore    28205
Rajasthan Royals               26242
Name: count, dtype: int64

In [45]:
data['bowling_team'].value_counts().head()

bowling_team
Mumbai Indians                 31505
Kolkata Knight Riders          29663
Chennai Super Kings            28576
Royal Challengers Bangalore    28358
Rajasthan Royals               26432
Name: count, dtype: int64

In [100]:
team_matches = data[['match_id', 'batting_team']].drop_duplicates()


match_counts = team_matches['batting_team'].value_counts().reset_index()
match_counts.columns = ['Team', 'Matches_Played']

fig = px.bar(
    match_counts.sort_values('Matches_Played', ascending=False),
    x='Team',
    y='Matches_Played',
    title='🏏 Number of Matches Played by Each Team',
    text='Matches_Played',
    color='Team',
    color_discrete_sequence=px.colors.qualitative.Set3
)
fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Team',
    yaxis_title='Number of Matches',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 600,
    showlegend = False
)

fig.show()

In [99]:
batter_matches = data[['match_id', 'batter']].drop_duplicates()


match_counts = batter_matches['batter'].value_counts().reset_index().head(20)
match_counts.columns = ['batter', 'Matches_Played']

fig = px.bar(
    match_counts.sort_values('Matches_Played', ascending=False),
    x='batter',
    y='Matches_Played',
    title='🏏 Number of Matches Played by Each batter',
    text='Matches_Played',
    color='batter',
    color_discrete_sequence=px.colors.qualitative.Antique
)
fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='batter',
    yaxis_title='Number of Matches',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 600,
    showlegend = False
)

fig.show()

In [101]:
bowler_matches = data[['match_id', 'bowler']].drop_duplicates()


match_counts = bowler_matches['bowler'].value_counts().reset_index().head(20)
match_counts.columns = ['bowler', 'Matches_Played']

fig = px.bar(
    match_counts.sort_values('Matches_Played', ascending=False),
    x='bowler',
    y='Matches_Played',
    title='🔴 Number of Matches Played by Each Bowler',

    text='Matches_Played',
    color='bowler',
    color_discrete_sequence=px.colors.qualitative.Set2_r
)
fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='bowler',
    yaxis_title='Number of Matches',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 600,
    showlegend = False
)

fig.show()

In [102]:
batter_runs = data.groupby('batter')['batsman_runs'].sum().reset_index()

batter_runs = batter_runs.sort_values(by='batsman_runs', ascending=False)
top_batters = batter_runs.head(10)

fig = px.bar(
    top_batters,
    x='batter',
    y='batsman_runs',
    title='🏏 Top 10 Highest Run Scorers',
    text='batsman_runs',
    color='batter',
    color_discrete_sequence=px.colors.qualitative.Vivid
)

fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Batter',
    yaxis_title='Total Runs Scored',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 550,
    showlegend = False
)

fig.show()

In [103]:
bowler_wicket = data.groupby('bowler')['is_wicket'].sum().reset_index()

bowler_wicket = bowler_wicket.sort_values(by='is_wicket', ascending=False)
top_bowlers = bowler_wicket.head(10)

fig = px.bar(
    top_bowlers,
    x='bowler',
    y='is_wicket',
    title='🏏🔴 Top 10 Highest Wicket Taker',
    text='is_wicket',
    color='bowler',
    color_discrete_sequence=px.colors.qualitative.Prism_r
)

fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='bowler',
    yaxis_title='Total Wicket Taken',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 550,
    showlegend = False
)

fig.show()

In [104]:
fielder_counts = data['fielder'].value_counts().reset_index()
fielder_counts.columns = ['fielder', 'appearance_count']

top_fielders = fielder_counts.head(20)

fig = px.bar(
    top_fielders,
    x='fielder',
    y='appearance_count',
    title='🧤 Top 10 Most Involved Fielders (Appearances in Dismissals)',
    text='appearance_count',
    color='fielder',
    color_discrete_sequence=px.colors.qualitative.Pastel1
)

fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Fielder',
    yaxis_title='Number of Appearances',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height=510,
    showlegend = False
)

fig.show()

In [105]:
sixes = data[data['batsman_runs'] == 6]

sixes_count = sixes['batter'].value_counts().reset_index()
sixes_count.columns = ['batter', 'sixes']

top_sixers = sixes_count.head(10)

fig = px.bar(
    top_sixers,
    x='batter',
    y='sixes',
    title='🔥 Top 10 Six Hitters',
    text='sixes',
    color='batter',
    color_discrete_sequence=px.colors.qualitative.Pastel
)

fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Batter',
    yaxis_title='Number of Sixes',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 520,
    showlegend = False
)
fig.show()

In [106]:
fours = data[data['batsman_runs'] == 4]

fours_count = fours['batter'].value_counts().reset_index()
fours_count.columns = ['batter', 'fours']

top_four_hitters = fours_count.head(10)

fig = px.bar(
    top_four_hitters,
    x='batter',
    y='fours',
    title= '4️⃣Top 10 Four Hitters',
    text='fours',
    color='batter',
    color_discrete_sequence=px.colors.qualitative.Pastel
)

fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Batter',
    yaxis_title='Number of Fours',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height = 550,
    showlegend = False
)

fig.show()

In [108]:
balls_faced = data['batter'].value_counts().reset_index()
balls_faced.columns = ['batter', 'balls_faced']

top_batters = balls_faced.head(10)

fig_bat = px.bar(
    top_batters,
    x='batter',
    y='balls_faced',
    title='🏏 Top 10 Batters by Balls Faced',
    text='balls_faced',
    color='batter',
    color_discrete_sequence=px.colors.qualitative.Vivid
)
fig_bat.update_traces(textposition='outside')
fig_bat.update_layout(
    xaxis_title='Batter',
    yaxis_title='Balls Faced',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height=500,
    showlegend = False
)
fig_bat.show()

balls_bowled = data['bowler'].value_counts().reset_index()
balls_bowled.columns = ['bowler', 'balls_bowled']

top_bowlers = balls_bowled.head(10)

fig_bowl = px.bar(
    top_bowlers,
    x='bowler',
    y='balls_bowled',
    title='🎯 Top 10 Bowlers by Balls Bowled',
    text='balls_bowled',
    color='bowler',
    color_discrete_sequence=px.colors.qualitative.Bold
)
fig_bowl.update_traces(textposition='outside')
fig_bowl.update_layout(
    xaxis_title='Bowler',
    yaxis_title='Balls Bowled',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height=550,
    showlegend = False
)
fig_bowl.show()

In [119]:
# ----------- Strike Rate Calculation --------
# Total runs scored by each batter
runs_by_batter = data.groupby('batter')['batsman_runs'].sum()

balls_faced = data['batter'].value_counts()

strike_rate_df = pd.DataFrame({
    'runs': runs_by_batter,
    'balls_faced': balls_faced
}).reset_index()

strike_rate_df['strike_rate'] = (strike_rate_df['runs'] / strike_rate_df['balls_faced']) * 100

top_sr = strike_rate_df[strike_rate_df['balls_faced'] >= 50].sort_values(by='strike_rate', ascending=False).head(10)

fig_sr = px.bar(
    top_sr,
    x='batter',
    y='strike_rate',
    title='⚡ Top 10 Batters by Strike Rate',
    text=top_sr['strike_rate'].round(2),
    color='batter',
    color_discrete_sequence=px.colors.qualitative.Vivid
)
fig_sr.update_traces(textposition='outside')
fig_sr.update_layout(
    xaxis_title='Batter',
    yaxis_title='Strike Rate',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height=500
)
fig_sr.show()

# -------- Runs Given by Bowler per Over ----------

min_balls_bowled = 18 * 6  # minimum bowled balls (18 overs)

bowler_stats_filtered = bowler_stats[bowler_stats['balls_bowled'] >= min_balls_bowled]

top_bowlers = bowler_stats_filtered.sort_values(by='runs_per_over').head(10)

fig_bowl = px.bar(
    top_bowlers,
    x='bowler',
    y='runs_per_over',
    title=f'🎯 Top 10 Economical Bowlers (Min {min_balls_bowled // 6} Overs)',
    text=top_bowlers['runs_per_over'].round(2),
    color='bowler',
    color_discrete_sequence=px.colors.qualitative.Pastel
)
fig_bowl.update_traces(textposition='outside')
fig_bowl.update_layout(
    xaxis_title='Bowler',
    yaxis_title='Runs per Over',
    xaxis={'categoryorder': 'total ascending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height=500
)
fig_bowl.show()





In [ ]:
wickets_data = data[data['is_wicket'] == 1]  

wickets_by_bowler_batter = wickets_data.groupby(['batter', 'bowler']).size().reset_index(name='wicket_count')

top_bowler_per_batter = wickets_by_bowler_batter.loc[wickets_by_bowler_batter.groupby('batter')['wicket_count'].idxmax()]

top_bowler_per_batter_sorted = top_bowler_per_batter.sort_values(by='wicket_count', ascending=False)

top_10 = top_bowler_per_batter_sorted.head(20)

fig = px.bar(
    top_10,
    x='batter',
    y='wicket_count',
    color='bowler',
    title='🔥 Top Bowlers with Most Wickets by Batter',
    text='wicket_count',
    color_discrete_sequence=px.colors.qualitative.Pastel1
)
fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Batter',
    yaxis_title='Wickets Taken',
    height=500,
    showlegend = False
)
fig.show()

In [123]:
fours_data = data[data['batsman_runs'] == 4]
fours_count = fours_data.groupby(['batter', 'bowler']).size().reset_index(name='fours_count')
top_fours = fours_count.sort_values(by='fours_count', ascending=False).head(10)

fig_fours = px.bar(
    top_fours,
    x='batter',
    y='fours_count',
    color='bowler',
    title='🔥 Top 10 Bowler-Batter Combos by Fours Conceded',
    text='fours_count',
    # barmode='group'
)
fig_fours.update_traces(textposition='outside')
fig_fours.update_layout(
    xaxis_title='Batter',
    yaxis_title='Number of Fours',
    height=500,
    showlegend = False
)
fig_fours.show()

### sixes-------------

sixes_data = data[data['batsman_runs'] == 6]
sixes_count = sixes_data.groupby(['batter', 'bowler']).size().reset_index(name='sixes_count')
top_sixes = sixes_count.sort_values(by='sixes_count', ascending=False).head(10)

fig_sixes = px.bar(
    top_sixes,
    x='batter',
    y='sixes_count',
    color='bowler',
    title='🔥 Top 10 Bowler-Batter Combos by Sixes Conceded',
    text='sixes_count',
    # barmode='group'
)
fig_sixes.update_traces(textposition='outside')
fig_sixes.update_layout(
    xaxis_title='Batter',
    yaxis_title='Number of Sixes',
    height=500,
    showlegend = False
)

fig_sixes.show()


In [124]:
data

match_id  inning           batting_team                 bowling_team  \
0         335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
1         335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
2         335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
3         335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
4         335982       1  Kolkata Knight Riders  Royal Challengers Bangalore   
...          ...     ...                    ...                          ...   
260915   1426312       2  Kolkata Knight Riders          Sunrisers Hyderabad   
260916   1426312       2  Kolkata Knight Riders          Sunrisers Hyderabad   
260917   1426312       2  Kolkata Knight Riders          Sunrisers Hyderabad   
260918   1426312       2  Kolkata Knight Riders          Sunrisers Hyderabad   
260919   1426312       2  Kolkata Knight Riders          Sunrisers Hyderabad   

        over  ball       batter         bowler  non_striker  batsman_runs  \
0          0     1   SC Ganguly        P Kumar  BB McCullum             0   
1          0     2  BB McCullum        P Kumar   SC Ganguly             0   
2          0     3  BB McCullum        P Kumar   SC Ganguly             0   
3          0     4  BB McCullum        P Kumar   SC Ganguly             0   
4          0     5  BB McCullum        P Kumar   SC Ganguly             0   
...      ...   ...          ...            ...          ...           ...   
260915     9     5      SS Iyer     AK Markram      VR Iyer             1   
260916     9     6      VR Iyer     AK Markram      SS Iyer             1   
260917    10     1      VR Iyer  Shahbaz Ahmed      SS Iyer             1   
260918    10     2      SS Iyer  Shahbaz Ahmed      VR Iyer             1   
260919    10     3      VR Iyer  Shahbaz Ahmed      SS Iyer             1   

        extra_runs  total_runs extras_type  is_wicket player_dismissed  \
0                1           1     legbyes          0              NaN   
1                0           0         NaN          0              NaN   
2                1           1       wides          0              NaN   
3                0           0         NaN          0              NaN   
4                0           0         NaN          0              NaN   
...            ...         ...         ...        ...              ...   
260915           0           1         NaN          0              NaN   
260916           0           1         NaN          0              NaN   
260917           0           1         NaN          0              NaN   
260918           0           1         NaN          0              NaN   
260919           0           1         NaN          0              NaN   

       dismissal_kind fielder  
0                 NaN     NaN  
1                 NaN     NaN  
2                 NaN     NaN  
3                 NaN     NaN  
4                 NaN     NaN  
...               ...     ...  
260915            NaN     NaN  
260916            NaN     NaN  
260917            NaN     NaN  
260918            NaN     NaN  
260919            NaN     NaN  

[260920 rows x 17 columns]

In [134]:
team_runs = data.groupby('batting_team')['total_runs'].sum().reset_index(name='total_runs')

team_matches = data.groupby('batting_team')['match_id'].nunique().reset_index(name='matches_played')

team_wickets = data[data['is_wicket'] == 1].groupby('batting_team')['is_wicket'].count().reset_index(name='wickets_lost')

team_balls = data.groupby('batting_team').size().reset_index(name='balls_faced')
team_balls['overs_faced'] = team_balls['balls_faced'] / 6

team_stats = team_runs.merge(team_matches, on='batting_team') \
                      .merge(team_wickets, on='batting_team') \
                      .merge(team_balls[['batting_team', 'overs_faced']], on='batting_team')

team_stats['run_rate'] = team_stats['total_runs'] / team_stats['overs_faced']
team_stats['wickets_per_match'] = team_stats['wickets_lost'] / team_stats['matches_played']

team_stats['score'] = (team_stats['run_rate'] * 2) - team_stats['wickets_per_match'] + team_stats['matches_played'] * 0.1

best_teams = team_stats.sort_values(by='score', ascending=False).reset_index(drop=True)


In [136]:
best_teams

batting_team  total_runs  matches_played  wickets_lost  \
0                Mumbai Indians       42176             261          1573   
1         Kolkata Knight Riders       39331             251          1491   
2           Chennai Super Kings       38629             237          1245   
3   Royal Challengers Bangalore       37692             240          1384   
4              Rajasthan Royals       34747             220          1312   
5               Kings XI Punjab       30064             190          1158   
6           Sunrisers Hyderabad       29071             182          1058   
7              Delhi Daredevils       24296             161           952   
8                Delhi Capitals       14900              91           570   
9               Deccan Chargers       11463              75           484   
10               Gujarat Titans        7757              45           247   
11                 Punjab Kings        9536              56           371   
12         Lucknow Super Giants        7510              44           276   
13  Royal Challengers Bengaluru        2930              15            99   
14                Gujarat Lions        4862              30           188   
15      Rising Pune Supergiants        2063              14            68   
16                Pune Warriors        6358              45           298   
17       Rising Pune Supergiant        2470              16            90   
18         Kochi Tuskers Kerala        1901              14            86   

    overs_faced  run_rate  wickets_per_match      score  
0   5239.500000  8.049623           6.026820  36.172426  
1   4919.000000  7.995731           5.940239  35.151223  
2   4775.166667  8.089561           5.253165  34.625957  
3   4700.833333  8.018153           5.766667  34.269639  
4   4373.666667  7.944593           5.963636  31.925549  
5   3774.333333  7.965380           6.094737  28.836024  
6   3640.500000  7.985442           5.813187  28.357696  
7   3131.000000  7.759821           5.913043  25.706599  
8   1824.333333  8.167367           6.263736  19.170998  
9   1505.666667  7.613239           6.453333  16.273144  
10   915.666667  8.471423           5.488889  15.953958  
11  1138.833333  8.373482           6.625000  15.721963  
12   900.000000  8.344444           6.272727  14.816162  
13   303.000000  9.669967           6.600000  14.239934  
14   594.333333  8.180595           6.266667  13.094522  
15   263.333333  7.834177           4.857143  12.211212  
16   907.166667  7.008635           6.622222  11.895048  
17   316.666667  7.800000           5.625000  11.575000  
18   263.666667  7.209861           6.142857   9.676865

In [ ]:
top_teams = best_teams.head(10)

fig = px.bar(
    top_teams,
    x='batting_team',
    y='score',
    text=top_teams['score'].round(2),
    color='batting_team',
    color_discrete_sequence=px.colors.qualitative.Bold,
    title='🏆 Top 10 Teams Based on Overall Batting Performance Score'
)

fig.update_traces(textposition='outside')
fig.update_layout(
    xaxis_title='Team',
    yaxis_title='Performance Score',
    xaxis={'categoryorder': 'total descending'},
    uniformtext_minsize=8,
    uniformtext_mode='hide',
    height=500
)
fig.show()

top payer by six/four+runs+filding+strickrate+less wicket = score sort by score
best team by all over + lowest defend + highest run + no. of match playes + all out